draft-ietf-quic-qpack.md

```diff
@@ -1304,54 +1304,62 @@ are registered in the "HTTP/3 Error Code" registry established in {{HTTP3}}.
 # Sample One Pass Encoding Algorithm
 
 Pseudo-code for single pass encoding, excluding handling of duplicates,
-non-blocking mode, and reference tracking.
+non-blocking mode, available encoder stream flow control and reference tracking.
 
 ~~~
-baseIndex = dynamicTable.baseIndex
-largestReference = 0
+base = dynamicTable.getInsertCount()
+requiredInsertCount = 0
 for header in headers:
-  staticIdx = staticTable.getIndex(header)
-  if staticIdx:
-    encodeIndexReference(streamBuffer, staticIdx)
+  staticIndex = staticTable.getIndex(header)
+  if staticIndex is not None:
+    encodeIndexReference(streamBuffer, staticIndex)
     continue
 
-  dynamicIdx = dynamicTable.getIndex(header)
-  if !dynamicIdx:
+  dynamicIndex = dynamicTable.getIndex(header)
+  if dynamicIndex is None:
     # No matching entry.  Either insert+index or encode literal
-    nameIdx = getNameIndex(header)
+
+    # getNameIndex attempts to find an index for this header's name.
+    # If one exists, it returns the name's (absolute) index and a
+    # boolean indicating which table has the name.  If the name
+    # can't be found, nameIndex is None
+    nameIndex, isStaticName = getNameIndex(header.name)
     if shouldIndex(header) and dynamicTable.canIndex(header):
-      encodeLiteralWithIncrementalIndex(controlBuffer, nameIdx,
-                                        header)
+      encodeInsert(encoderBuffer, nameIndex, isStaticName, header)
       dynamicTable.add(header)
-      dynamicIdx = dynamicTable.baseIndex
+      dynamicIndex = dynamicTable.getInsertCount()
```

Comment:
I can't suggest here, but I think that you just want this:

```python
        dynamicIndex = dynamicTable.add(header)
```

Otherwise you are breaking encapsulation in a small way: relying on the insert count for getting the index of the thing you just added.